nifi-nar-bundles/nifi-standard-services/nifi-oauth2-provider-bundle/nifi-oauth2-provider-service/src/main/java/org/apache/nifi/oauth2/PasswordBasedOauth2TokenProvider.java

```diff
@@ -0,0 +1,230 @@
+        .build();
+
+    public static final PropertyDescriptor CLIENT_SECRET = new PropertyDescriptor.Builder()
+        .name("client-secret")
+        .displayName("Client secret")
+        .dependsOn(CLIENT_ID)
+        .required(true)
+        .sensitive(true)
+        .addValidator(StandardValidators.NON_BLANK_VALIDATOR)
+        .build();
+
+    public static final PropertyDescriptor SSL_CONTEXT = new PropertyDescriptor.Builder()
+        .name("ssl-context")
+        .displayName("SSL Context")
+        .addValidator(Validator.VALID)
+        .identifiesControllerService(SSLContextService.class)
+        .required(false)
+        .build();
+
+    private static final List<PropertyDescriptor> PROPERTIES = Collections.unmodifiableList(Arrays.asList(
+        AUTHORIZATION_SERVER_URL,
+        USERNAME,
+        PASSWORD,
+        CLIENT_ID,
+        CLIENT_SECRET,
+        SSL_CONTEXT
+    ));
+
+    private String authorizationServerUrl;
+    private OkHttpClient httpClient;
+
+    private String username;
+    private String password;
+    private String clientId;
+    private String clientSecret;
+
+    private volatile AccessDetails accessDetails;
+
+    @Override
+    public List<PropertyDescriptor> getSupportedPropertyDescriptors() {
+        return PROPERTIES;
+    }
+
+    @OnEnabled
+    public void onEnabled(ConfigurationContext context) {
+        authorizationServerUrl = context.getProperty(AUTHORIZATION_SERVER_URL).evaluateAttributeExpressions().getValue();
+
+        httpClient = createHttpClient(context);
+
+        username = context.getProperty(USERNAME).evaluateAttributeExpressions().getValue();
+        password = context.getProperty(PASSWORD).getValue();
+        clientId = context.getProperty(CLIENT_ID).evaluateAttributeExpressions().getValue();
+        clientSecret = context.getProperty(CLIENT_SECRET).getValue();
+    }
+
+    protected OkHttpClient createHttpClient(ConfigurationContext context) {
+        OkHttpClient.Builder clientBuilder = new OkHttpClient.Builder();
+
+        SSLContextService sslService = context.getProperty(SSL_CONTEXT).asControllerService(SSLContextService.class);
+        if (sslService != null) {
+            final X509TrustManager trustManager = sslService.createTrustManager();
+            SSLContext sslContext = sslService.createContext();
+            clientBuilder.sslSocketFactory(sslContext.getSocketFactory(), trustManager);
+        }
+
+        return clientBuilder.build();
+    }
+
+    @Override
+    public String getAccessToken() {
+        String accessToken;
+
+        if (this.accessDetails == null) {
+            acquireAuthorizationDetails();
+        } else if (this.accessDetails.isExpired()) {
+            try {
+                refreshAuthorizationDetails();
+            } catch (Exception e) {
+                getLogger().info("Couldn't refresh access token, probably refresh token has expired." +
```

Comment:
It seems like this should be a warning instead of an information message, as the general `Exception` catch could include a wide variety of problems. Recommend more precise error handling instead of the general message saying `probably refresh token has expired`.

Reply:
I think overall the current implementation is more optimal for the following reason:

This is a trial-and-error approach where the "error" is to be expected to happen when the refresh token expires.
Definitely wouldn't want a warning in that case.
Even if the issue is something else, we try to get a new token. If that succeeds we don't really care about why the refresh failed. If we can't get a new token, we report the exception as usual so we won't be losing anything as far as reporting meaningful error goes.

Reply:
At minimum it would be helpful to differentiate between HTTP response exceptions and other types of exceptions. Problems such as socket timeouts or connection failures should not result it attempting to acquire a new token, but should be propagated as errors. Including more well-defined error handling would make troubleshooting problems much easier.

Reply:
I'll update the error message and remove the kind of arbitrary "probably refresh token has expired" part as it serves little to no purpose.
Otherwise I tested all error scenarios and in all cases I see the issue properly reported.
The information provided is clear and lacks no detail.
Unless presented a concrete detailed and reproducible example that would prove otherwise I consider this level of error handling sufficient.